src/callbacks_step/time_series_dg2d.jl

```diff
@@ -68,6 +70,141 @@ function get_elements_by_coordinates!(element_ids, coordinates, mesh::TreeMesh,
+        P = surface_point .- point
+
+        # If the vector `P` is the zero vector then this `point` is at an element corner or
+        # on a surface. In this case the choice of a candidate element is ambiguous and
+        # we just use the first candidate. However, solutions might differ at discontinuous
+        # interfaces such that this choice may influence the result.
+        if sum(P .* P) < 500 * eps(eltype(point))
+            element_ids[index] = candidates[1]
+            continue
+        end
+
+        # Loop through all the element candidates until we find a vector from the barycenter
+        # to the surface that points in the same direction as the current `point` vector.
+        # This then gives us the correct element.
+        for element in 1:length(candidates)
+            bary_center = SVector(bary_centers[1, candidates[element]],
+                                  bary_centers[2, candidates[element]])
+            # Vector pointing from the barycenter toward the minimal `surface_point`
+            B = surface_point .- bary_center
+            if sum(P .* B) > zero(eltype(bary_center))
+                element_ids[index] = candidates[element]
+                break
+            end
+        end
+    end
+
+    return element_ids
+end
+
+# Use the available `node_coordinates` on each element to compute and save the barycenter.
+# In essence, the barycenter is like an average where all the x and y node coordinates are
+# summed and then we divide by the total number of degrees of freedom on the element, i.e.,
+# the value of `n^2` in two spatial dimensions.
+@inline function calc_bary_centers!(bary_centers, dg, cache)
+    n = nnodes(dg)
+    @views for element in eachelement(dg, cache)
+        bary_centers[1, element] = sum(cache.elements.node_coordinates[1, :, :,
+                                                                       element]) / n^2
+        bary_centers[2, element] = sum(cache.elements.node_coordinates[2, :, :,
+                                                                       element]) / n^2
+    end
+    return nothing
+end
+
+# Compute the shortest distance from a `point` to the surface of each element
+# using the available `node_coordinates`. Also return the index pair of this
+# minimum surface point location. We compute and store in `min_distance`
+# the squared norm to avoid computing computationally more expensive square roots.
+# Note! Could be made more accurate if the `node_coordinates` were super-sampled
+# and reinterpolated onto a higher polynomial degree before this computation.
+function calc_minimum_surface_distance(point, node_coordinates,
+                                       dg, mesh::UnstructuredMesh2D)
+    n = nnodes(dg)
+    min_distance = Inf * ones(eltype(mesh.corners), length(mesh))
+    indices = zeros(Int, length(mesh), 2)
+    for k in 1:length(mesh)
+        # used to ensure that only boundary points are used
+        on_surface = [false, false]
```

Comment:
This should avoid allocating lots of small vectors (since the compiler should be smart enough to realize that the `MVector` doesn't escape and can thus be allocated on the stack):

```suggestion
        on_surface = MVector(false, false)
```